Spark: Custom snapshot property from session configuration

cccs-jory opened the pull request:
Adds support to allow custom snapshot properties to be specified in the Spark session configuration. This allows users to add custom snapshot properties even when running spark SQL DML such as DELETE or MERGE INTO, which was not previously supported.

guykhazma commented:
Considering that the goal is to be able to set it dynamically for each query wouldn't it make more sense to see if there is a way to include it in the query itself instead of having it as a session configuration.
I don't know of  way to pass properties using the SQL syntax but I wonder whether the [hinting mechanism](https://spark.apache.org/docs/latest/sql-ref-syntax-qry-select-hints.html) in spark or something similar (if exists) can be used to communicate this kind of properties dynamically as part of the query.

cccs-jory commented:
> Considering that the goal is to be able to set it dynamically for each query

I don't think that's the goal necessarily. For example an org may create a spark session that executes DML on many tables, therefore creating many commits. That spark session could set this configuration and have its custom snapshot property applied on each table's commits. Setting it dynamically in Spark SQL could be a different task?

singhpk234 reviewed (commented):
Similar requirements did come up in the past : https://github.com/apache/iceberg/pull/4956 is this not sufficient ? or you want this for specifically with the SQLConfs ? 

Can you please elaborate your use case : 
is the concern this is not supported for DELETE / MERGE and only INSERT ? if yes we should fix that.

cccs-jory commented:
> Similar requirements did come up in the past : #4956 is this not sufficient ? or you want this for specifically with the SQLConfs ?
> 
> Can you please elaborate your use case : is the concern this is not supported for DELETE / MERGE and only INSERT ? if yes we should fix that.

`withCommitProperties` is only accessible via Java. If teams run PySpark then there is no way to access it (outside of using `py4j`).

RussellSpitzer commented:
Please only target one Spark Version in this PR for easier reviewing, we can do a fast followup that changes earlier versions after we merge this.

RussellSpitzer commented on spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/TestSparkWriteConf.java:
```diff
@@ -486,6 +486,36 @@ public void testDeleteFileWriteConf() {
+  @Test
+  public void testExtraSnapshotMetadataReflectsSessionConfig() {
+    withSQLConf(
+        ImmutableMap.of("spark.sql.iceberg.snapshot-property.test-key", "test-value"),
```
Trivial nit: "test-value" -> "session-value" so it mimics the test below. This is the tiniest of nits so feel free to ignore.

RussellSpitzer reviewed (approved):
I have a tiny nit, and since these changes are so small I don't think we need to split apart the PR into different versions.

RussellSpitzer commented:
Looks like there is a build issue @cccs-jory, please update and let me know when to run tests again :)

cccs-jory commented:
@RussellSpitzer  I removed the feature on 3.4 and addressed the nitpick, however I may need some more eyes to fix the build issue.

Essentially gradlew is failing when compiling on 3.5 because it can't find Spark's `RuntimeConfig.getAll()`, however when I look at the Javadocs it is clearly there and exists. Even if I switch out `sessionConf.getAll()` for `spark.conf().getAll()`, it still fails.

RussellSpitzer commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java:
```diff
@@ -252,6 +252,15 @@ public long targetDeleteFileSize() {
     Map<String, String> extraSnapshotMetadata = Maps.newHashMap();
 
+    // Check session configuration for properties with SNAPSHOT_PROPERTY_PREFIX
+    sessionConf.getAll().forEach((key, value) -> {
```
The issue here is "getAll()" is returning a Scala Map and not a Java Map

RussellSpitzer replied:
You could do this
```java
    // Check session configuration for properties with SNAPSHOT_PROPERTY_PREFIX
    JavaConverters.mapAsJavaMap(sessionConf.getAll()).forEach((key, value) -> {
      if (key.startsWith(SparkSQLProperties.SNAPSHOT_PROPERTY_PREFIX)) {
        extraSnapshotMetadata.put(
            key.substring(SparkSQLProperties.SNAPSHOT_PROPERTY_PREFIX.length()), value);
      }
    });
```

cccs-jory replied:
great, should be fixed now!

anuragmantri commented:
I think this is still relevant and not stale. The default version of Spark changed to 4.0 so we may want to move the changes to that branch. @cccs-jory, could you address comments and update the PR?

RussellSpitzer reviewed (commented):
Any follow up on this? Looks like we are just waiting on the change from @singhpk234 ?

owen6314 commented:
Any update on this PR? We've found this custom snapshot property support from session configuration helpful for our use cases, wondering if anything is blocking the merge.
@cccs-jory
